### PR TITLE
hal: apollo4p: usb: add UDC support 

### PR DESCRIPTION
**Pull request by zeonchew**
Added changes needed to support UDC driver including: 
1. `CONFIG_AMBIQ_HAL_USE_USB` flag to decide whether to include USB HAL for compilation
1. `AM_HAL_USB_CTRL_XFR_WAIT_STATUS_ACK_ZLP_FROM_STACK` flag to control EP0 control transfer behavior
1. `am_hal_usb_ep_state_reset` API to reset EP state by UDC driver
1.  Handle condition where OUT packet receive completed before UDC assign endpoint buffer to the endpoint.
1.  Refined suspend/resume event handling, to avoid false suspend event during bus reset.

### CODE REVIEW

**Comment by zeonchew** [comment]
Hi @RichardSWheatley , @aaronyegx , @AlessandroLuo , 

can you please help to review this PR? 
Also, it seems that I have no permission to assign reviewer here for some reason, do you know why?

**Comment by RichardSWheatley** [comment]
All of these code changes need to be made to Ambiqsuite stable as well. This is to keep the files the same and not need to track differences.

The process is to add everything to Ambiqsuite and then bring to Zephyr.

File: mcu/apollo4p/hal/am_hal_usb.c
Comment (RichardSWheatley) [inline mcu/apollo4p/hal/am_hal_usb.c:1961]: For ifdefs, Ambiqsuite typically has these aligned to the left.

**Review by RichardSWheatley — APPROVED** [review]
Just make sure these get formatted for Ambiqsuite and make sure they are already in Ambiqsuite/

**Comment by RichardSWheatley** [comment]
> Hi @RichardSWheatley , @aaronyegx , @AlessandroLuo , 
> 
> 
> 
> can you please help to review this PR? 
> 
> Also, it seems that I have no permission to assign reviewer here for some reason, do you know why? 

This is Ambiq and not Zephyr.
You just need to be added.

**Comment by zeonchew** [comment]
Hi @RichardSWheatley, 

Thanks for the review. I have made the coding style changes in commit: [commit 5d38eda52d1](https://github.com/zephyrproject-rtos/hal_ambiq/pull/32/commits/5d38eda52d10ddd111b820e38acc0944ac925502). 

A PR is also raised for ambiqsuite for the same changes in this PR. Will follow the flow you mentioned, and wait for the PR to be merged into ambiqsuite before merging this PR.

**Review by AlessandroLuo — APPROVED** [review]

**Pull request merged**
